src/AbstractLinesBeforeNamespaceFixer.php

```diff
@@ -34,22 +35,66 @@
+        // and the opening token
+        $precedingNewlinesTotal = 0;
+        $precedingNewlinesInOpening = 0;
+        $openingToken = null;
+        for ($i = 1; $i <= 2; ++$i) {
+            if (isset($tokens[$index - $i])) {
+                $token = $tokens[$index - $i];
+                if ($token->isGivenKind(T_OPEN_TAG)) {
+                    $openingToken = $token;
+                    $precedingNewlinesInOpening = substr_count($token->getContent(), "\n");
+                    $precedingNewlinesTotal += $precedingNewlinesInOpening;
+                    break;
+                }
+                if (false === $token->isGivenKind(T_WHITESPACE)) {
+                    break;
+                }
+                $precedingNewlinesTotal += substr_count($token->getContent(), "\n");
             }
         }
 
-        $previousIndex = $index - 1;
-        $previous = $tokens[$previousIndex];
-        if ($previous->isWhitespace()) {
+        if ($expected !== $precedingNewlinesTotal) {
+            $previousIndex = $index - 1;
+            $previous = $tokens[$previousIndex];
             if (0 === $expected) {
-                $tokens->clearAt($previousIndex);
-            } elseif (substr_count($previous->getContent(), "\n") !== $expected) {
-                $tokens[$previousIndex] = new Token(array(T_WHITESPACE, str_repeat("\n", $expected)));
+                // Remove all the previous new lines
+                if ($previous->isWhitespace()) {
+                    $tokens->clearAt($previousIndex);
+                }
+                // Remove new lines in opening token
+                if (0 < $precedingNewlinesInOpening) {
+                    $openingToken->setContent(rtrim($openingToken->getContent()).' ');
+                }
+            } else {
+                $lineEnding = $this->whitespacesConfig->getLineEnding();
+                if (null !== $openingToken) {
+                    $content = rtrim($openingToken->getContent());
+                    if (0 === $precedingNewlinesInOpening) {
+                        // We have an opening tag without new lines: add a new line there
+                        $openingToken->setContent($content.$lineEnding);
+                        ++$precedingNewlinesInOpening;
+                    } else {
+                        // Use the configured line ending for the PHP opening tag
+                        $openingToken->setContent($content.str_repeat($lineEnding, $precedingNewlinesInOpening));
+                    }
+                }
+                $newlinesForWhitespaceToken = $expected - $precedingNewlinesInOpening;
+                if ($previous->isWhitespace()) {
+                    if (0 === $newlinesForWhitespaceToken) {
+                        // We have all the needed new lines in the opening tag
+                        // Let's remove the previous token containing extra new lines
+                        $tokens->clearAt($previousIndex);
+                    } else {
```

Comment:
There are too much nested `if/else` IMO. Isn't it possible to add some early returns?